integrations/clickup/clickup/client.py

```diff
@@ -0,0 +1,147 @@
+            json={
+                "endpoint": webhook_target_app_host,
+                "events": WEBHOOK_EVENTS,
+            },
+        )
+        webhook_create_response.raise_for_status()
+        webhook_create = webhook_create_response.json()
+        logger.info(
+            f"Ocean real time reporting clickup webhook created "
+            f"[ID: {webhook_create['id']}, Team ID: {team_id}]"
+        )
+
+    async def get_teams(self, params: dict[str, Any] = {}) -> list[dict[str, Any]]:
+        teams_response = await self.client.get(
+            f"{self.clickup_url}/team", params=params
+        )
+        teams_response.raise_for_status()
+        teams = teams_response.json()["teams"]
+        return teams
+
+    async def _get_spaces(
+        self, team_id: str, params: dict[str, Any] = {}
+    ) -> list[dict[str, Any]]:
+        spaces_response = await self.client.get(
+            f"{self.clickup_url}/team/{team_id}/space", params=params
+        )
+        spaces_response.raise_for_status()
+        return spaces_response.json()["spaces"]
+
+    async def get_projects(self, params: dict[str, Any] = {}) -> list[dict[str, Any]]:
+        # getting all teams so as to retrieve the spaces within each team, then using their
+        # ids to get the projects (lists) within each space
+        projects: list[dict[str, Any]] = []
+        for team in await self.get_teams(params):
+            spaces = await self._get_spaces(team["id"], params)
+
+            for space in spaces:
+                projects_response = await self.client.get(
+                    f"{self.clickup_url}/space/{space['id']}/list", params=params
+                )
+                projects_response.raise_for_status()
+                # because the port-app-config uses the team ID to relate the projects to the teams
+                # we add the team ID to each project
+                # also, the clickup returns the datetime as a timestamp, so we convert it to a
+                # datetime object
+                projects.extend(
+                    map(
+                        lambda project: {
+                            **project,
+                            "__team_id": team["id"],
+                        },
+                        projects_response.json()["lists"],
+                    )
+                )
+        return projects
+
+    async def get_single_project(self, project_id: str) -> dict[str, Any]:
+        # Clickup does not provide a direct way to get the team ID from a project (list), so instead
+        # of directly getting the project, we get all the projects and filter by the project ID
+        # using the result of the get_projects method which includes the team ID
+        project = next(
+            (
+                project
+                for project in await self.get_projects()
+                if project["id"] == project_id
+            ),
+            None,
+        )
```

Comment:
as although it is nice it re-uses the same code as in the resync method, it is quite an not efficient way to resolve and get the project, as it is very costly to get list all the teams and then all the spaces and then all the projects just to get a single project.

Reply:
Yes, that's true. I'll think of a different approach